Framework/Algorithms/src/ConvertUnits.cpp

```diff
@@ -686,27 +675,24 @@ API::MatrixWorkspace_sptr ConvertUnits::removeUnphysicalBins(
                                                  maxBins - 1);
     // Next, loop again copying in the correct range for each spectrum
     for (int64_t j = 0; j < int64_t(numSpec); ++j) {
-      const MantidVec &X = workspace->readX(j);
-      const MantidVec &Y = workspace->readY(j);
-      const MantidVec &E = workspace->readE(j);
-      MantidVec &Xnew = result->dataX(j);
-      MantidVec &Ynew = result->dataY(j);
-      MantidVec &Enew = result->dataE(j);
-      int k;
-      for (k = 0; k < lastBins[j] - 1; ++k) {
-        Xnew[k] = X[k];
-        Ynew[k] = Y[k];
-        Enew[k] = E[k];
-      }
-      Xnew[k] = X[k];
-      ++k;
-      // If necessary, add on some fake values to the end of the X array (Y&E
-      // will be zero)
-      if (k < maxBins) {
-        for (int l = k; l < maxBins; ++l) {
-          Xnew[l] = X[k] + 1 + l - k;
-        }
-      }
+      auto edges = workspace->binEdges(j);
+
+      result->mutableX(j).assign(edges.cbegin(), edges.cbegin() + lastBins[j]);
+
+      // If the entire X range is not covered, generate fake values.
+      auto l = lastBins[j];
+
+      std::generate(result->mutableX(j).begin() + l, result->mutableX(j).end(),
+                    [=]() mutable {
+                      ++l;
+                      return result->mutableX(j)[lastBins[j]] + 1 +
+                             static_cast<double>(l - lastBins[j]);
+                    });
```

Comment:
I think something here is messed up with the indices. You are always using `result->mutableX(j)[lastBins[j]]` as base value, but it used to be `X[k]`.